### ISSUE TITLE
Ongoing call not disconnected when a third call is rejected

### ISSUE THREAD

**Issue by nicolettemanas** [twilio/voice-quickstart-objc#48]
### Description
Hi Twilio team, 

I am using **2.0.0-beta13** and I am experiencing a regression bug (last fixed in **2.0.0-beta7**) where an ongoing call is not disconnected if a third call is rejected. 

`callDidDisconnect:` is not called after the other user disconnects from an ongoing call.

### Steps to Reproduce
1. Establish a call between User1 and User2
2. User3 tries to call User1
3. User1 rejects the call invite as there is already an ongoing call
4. Voice call between User1 and User2 continues
5. User1 drops the call

Results: `callDidDisconnect:` is not called in User2

I tested the scenario using 2.0.0-beta12; the calls are disconnected fine, but [the ongoing voice call is destroyed](https://github.com/twilio/voice-quickstart-objc/issues/44)

#### Code

I added `[callInvite reject];` in the quickstart app upon receiving a call invite during an ongoing call. I included the verbose logs [here](https://github.com/twilio/voice-quickstartobjc/files/1169339/twilio-logs.txt).

#### Voice iOS SDK
2.0.0-beta13 via CocoaPods

#### Xcode
8.3.2

#### iOS Versions and iOS Devices
Tested using iPhone6s (10.1.1), iPad Air (10.1.1), iPad Mini (9.3.5)

**Comment by ceaglest** [comment]
Hi @nicolettemanas,

Is this a duplicate of the ticket #44 that you filed, but reproduced with beta13 instead of beta12?

Based on the title alone - the ongoing call should **not** be disconnected when a third call is rejected. Lets look at the reproduction steps.

> 1. Establish a call between User1 and User2
> 2. User3 tries to call User1
> 3. User1 rejects the call invite as there is already an ongoing call
> 4. Voice call between User1 and User2 continues
> 5. User1 drops the call

What does the 5th step mean, is this the lack of audio that you reported earlier? The logs give a 404, can you re-upload them?

> Results: callDidDisconnect: is not called in User2

Since the active call should not end I would not expect `callDidDisconnect:` to be raised to User1 or User2.

If this is a duplicate I would prefer to add this information to the original ticket and continue to discuss it there.

Thanks,
Chris Eagleston

**Comment by nicolettemanas** [comment]
Hi @ceaglest 

Sorry for the confusion, this is not a duplicate of #44 , the 5th step actually means User1 _intentionally_ ending the call and therefore, `callDidDisconnect:` _should_ be called in User2 (but is not being called) using **2.0.0-beta13**.

This issue was filed and fixed in 2.0.0-beta7, I'd link the github issue but it seems that issues in the **voice-callkit-quickstart-objc** project were removed. Please refer to the fix `CLIENT-3513` in the [changelog](https://www.twilio.com/docs/api/voice-sdk/ios/changelog).

I reploaded the logs [here](https://github.com/twilio/voice-quickstart-objc/files/1171616/twilio-logs.txt).

Thanks,
Jean

**Comment by ceaglest** [comment]
Hi,

Thank you for the clarification! 

Sorry for the confusion from my side as well, we deactivated the issues for `voice-callkit-quickstart-objc` because that repository is deprecated.

I will check with @bchen-twilio on the history for this one, and get back to you.

Thanks,
Chris

**Comment by nicolettemanas** [comment]
Hello,

Thank you for the update! Let me know if you need further details.

Thanks!

**Comment by nicolettemanas** [comment]
Hi Twilio team, 

Any updates on this? 

Regards,
Jean

**Comment by ceaglest** [comment]
Hi Jean,

I'm sorry for the delayed response. I'm working on reproducing the issue this afternoon, but the steps you provided are perfectly clear.

One thing that is concerning based upon your logs and my local tests is that we may be over-releasing the signaling transport upon User 1 rejecting the invite from User 3. As a part of this investigation I am adding integration test coverage for the scenario you describe, as our current test cases in this area are not a perfect match.

To make sure we are abundantly clear on the modifications to the QS I posted a [branch](https://github.com/twilio/voice-quickstart-objc/tree/issue/48-ongoing-call-disconnect-missed) with the reject modification that you described.

I will give another update in the next few hours.

Thanks,
Chris Eagleston

**Comment by nicolettemanas** [comment]
Hi Chris,

It's pretty much the same changes, except for the parts where we had to modify the parameters in TwilioVoice's `-call:params:delegate` to establish a call between users. 

Were you able to reproduce this in beta13?

I'll try providing the verbose logs for the three users this afternoon.

Hope this helps,
Jean

**Comment by ceaglest** [comment]
Hi Jean,

I wasn't able to reproduce in beta13 yet, but my User2 and User3 were JavaScript clients. I will re-test with all iOS clients, and am also adding integration tests so we can catch this automatically in the future.

Logs for all 3 Clients would be really helpful if you can provide them.

Thanks,
Chris

**Comment by nicolettemanas** [comment]
Sure, I'll upload them here when I get my hands on the devices. Thanks!

**Comment by nicolettemanas** [comment]
Hi Chris,

Sorry for the late revert. Here are the logs for [user1.txt](https://github.com/twilio/voice-quickstart-objc/files/1207061/user1.txt), [user2.txt](https://github.com/twilio/voice-quickstart-objc/files/1207062/user2.txt), and [user3.txt](https://github.com/twilio/voice-quickstart-objc/files/1207060/user3.txt).

Some extra findings that might help: 
- It matters which user rejects the third party call and which user disconnects from the ongoing call 
- In an ongoing call between user1 and user2, if user2 gets and rejects the call from user3, the issue only happens if user1 is the one who drops the current call afterwards

In short, the issue only occurs if the user who didn't receive the third party call is the one who disconnects from the ongoing call. 

As for the logs provided, user2 is the one who gets the third party call invite. 

**user1**: iPhone 6s | iOS 10.1.1 
**user2**: iPhone 6+ | iOS 11.1
**user3**: iPad mini | iOS 8.2

Hope this helps,
Jean

**Comment by bobiechen-twilio** [comment]
Hi @nicolettemanas 

It's me again.
After checking the logs of user1/2/3, I think this is totally a regression you reported and we fixed a while ago, though the steps you described are slightly differently from what I saw in the logs.

Could you confirm the steps:
1. launch app on two devices and register as "alice" and "bob"
2. client "alice" calls "bob"
3. while the call is connected and both clients are talking to each other, trigger an incoming call to "alice"
4. "alice" rejects the call
5. at this point call audio between "alice" and "bob" should still flow
6. "bob" hangs up

The result would be that the call on "bob" is disconnected but "alice" is not getting the disconnected callback.

Let us know if the steps are correct.
Thanks.

-bobie

**Comment by nicolettemanas** [comment]
Hi @bchen-twilio ,

Those are exactly the steps. User1 being "bob" and User2 being "alice". I'm not sure which parts of the logs are different, could you please elaborate? 

Thanks,
Jean

**Comment by bobiechen-twilio** [comment]
Hi @nicolettemanas 

I am sorry that my previous comment confused you - what I meant is originally in your description, `user1` is the one who rejects the invite while in an active call with `user2` but then I remembered and realized it is `user2` who should reject the call and not getting the disconnected event.

Anyway we've created another ticket for tracking this regression and will let you know once it is fixed.

Thanks
bobie

**Comment by nicolettemanas** [comment]
Hi Bobie,

I see. Sorry about that. It happens as long as the one who drops the call at the end is the user who didn't receive the third party call. 

Thanks for the update!
Jean

**Comment by ceaglest** [comment]
Hi Jean,

This bug has been assigned a ticket number CLIENT-3970, and the fix is undergoing internal testing. We expect that the ticket will land in `2.0.0-beta15`. As part of our efforts Bobie added a test case which consistently reproduced the missing disconnect event. Once we were able to reliably reproduce it I fixed the regression as part of some internal refactoring work.

Best,
Chris Eagleston

**Comment by bobiechen-twilio** [comment]
Hi @nicolettemanas 

Chris is out of office for a couple of days so I'd just take the liberty of replying to you.
We've just released the Voice SDK beta15 which has the fix for this issue. Could you update the Podfile and pull the latest SDK?

Thanks.
-bobie

**Comment by nicolettemanas** [comment]
Hi @bchen-twilio , @ceaglest 

Sorry for the late revert. Confirming that I am not experiencing this issue anymore in **beta15**. 
Thank you for the support and working with this issue. 😄 

Best regards,
Jean

**Comment by ceaglest** [comment]
Thanks Jean,

That's great to hear! I'm going to go ahead and close out this ticket now that beta15 is in the wild.

Best,
Chris Eagleston
